### PR TITLE
Populate cdn-definitions with various cdn-utils data [RHELDST-4797]

### PR DESCRIPTION
**Pull request by crungehottman**

### CODE REVIEW

File: tests/test_schema.py [hunk 1]
```diff
@@ -45,3 +45,80 @@ def test_relative_path_not_match_schema(data, schema):
...
     with pytest.raises(ValidationError):
         jsonschema.validate(data, schema)
+
+
+def test_empty_list(data, schema):
+    """Verify that YAML returns empty list."""
+    assert data["signing_keys_mappings"]["platform"]["middleware"]["ga_keys"] == []
+
+
+def test_invalid_arches(data, schema):
```
Comment (crungehottman) [inline tests/test_schema.py:55]: I only added tests to validate regexes. I wasn't sure that it was necessary to include tests that verify that the json schema won't allow additional properties that don't match the specified type, or pattern (if the type is a string), etc.

**Comment by crungehottman** [comment]
I included a note about the data that I chose (or didn't choose) to include here: https://projects.engineering.redhat.com/browse/RHELDST-4797?focusedCommentId=3087975&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-3087975

**Comment by crungehottman** [comment]
Also note that this PR contains https://github.com/release-engineering/cdn-definitions/pull/19, so please only review the most recent commit.

**Comment by crungehottman** [comment]
CC: @nathanegillett  @lipoja (I cannot add either of you as reviewers)

File: tests/test_schema.py [hunk 2]
```diff
@@ -45,3 +45,80 @@ def test_relative_path_not_match_schema(data, schema):
     data["rhui_alias"][0]["dest"] = "dest-path"
     with pytest.raises(ValidationError):
         jsonschema.validate(data, schema)
+
+
+def test_empty_list(data, schema):
+    """Verify that YAML returns empty list."""
```
Comment (nathanegillett) [inline tests/test_schema.py:51]: This doesn't seem necessary. We shouldn't need to test the functionality of a third-party library.

**Review by nathanegillett — DISMISSED** [review]

File: src/cdn_definitions/data.yaml [hunk 3]
```diff
@@ -92,3 +92,136 @@ origin_alias:
...
+  - "8.0"
+  tus:
+  - "6.7"
+  - "8.0"
+  e4s:
+  - "7.2"
+  - "8.0"
+
+
+# Maps an environment to it's current and latest dist and beta RHEL releasevers.
```
Comment (rohanpm) [inline src/cdn_definitions/data.yaml:121]:
```suggestion
# Maps an environment to its current and latest dist and beta RHEL releasevers.
```

File: src/cdn_definitions/schema.yaml [hunk 4]
```diff
@@ -11,7 +11,211 @@ $id: http://release-engineering.github.io/cdn-definitions/schema.json
...
+    type: string
+    pattern: "^[0-9]+\\.[0-9]+$"
+
+  major_version:
+    type: string
+    pattern: "^[0-9]+$"
+
+  signing_key:
+    type: string
+    pattern: "^(37036783|A5787476|FD431D51|F21541EB|37017186|897DA07A|DB42A60E|\\[\\])$"
```
Comment (rohanpm) [inline src/cdn_definitions/schema.yaml:33]: I think this one doesn't belong in the schema.  Could we just enforce the format instead and not make this an enum of values?

I don't like the idea that adding a new signing key requires updating the schema, or that one is unable to use cdn-definitions in their own test environment with their own personal signing key.

By the way, what's the meaning of ```[]``` as signing_key ?

File: src/cdn_definitions/schema.yaml [hunk 5]
```diff
@@ -34,6 +238,11 @@ definitions:
       $ref: "#/definitions/path_alias"
     uniqueItems: true
 
+  symlink_exception:
+    type: array
+    items:
+      type: string
+      pattern: "^\\/*"
```
Comment (rohanpm) [inline src/cdn_definitions/schema.yaml:245]: This pattern doesn't seem right to me, though perhaps the YAML layer is confusing me.

- Firstly, "/" isn't a special character in regexes, so isn't escaping it unnecessary?  Does dropping the "\\" make any difference?
- Secondly, "/*" means "zero or more instances of /" but this doesn't seem useful since it'll match against anything at all.

I guess the intent was to ensure it's an absolute path but I think this is not the correct pattern.  Examples:

```python
# Matches any random string
>>> re.search(r"\/*", 'quux')
<re.Match object; span=(0, 0), match=''>

# Backslash doesn't matter
>>> re.search(r"/*", 'quux')
<re.Match object; span=(0, 0), match=''>

# Maybe it was supposed to be this?
>>> re.search(r"/.*", 'quux')
```

File: src/cdn_definitions/data.yaml [hunk 6]
```diff
@@ -92,3 +92,136 @@ origin_alias:
 # permanent.
 - src: /origin/rpm
   dest: /origin/rpms
+
```
Comment (rohanpm) [inline src/cdn_definitions/data.yaml:95]: This seems to be a snapshot of real data, maybe cut down a little, but we know the real data is meant to come from cdn-definitions-private.

Is the idea then that the reference data is initially populated from a snapshot but then allowed to diverge over time?

I don't think it's a good idea to set it up that way for a couple of reasons:

- If this data looks like real data then people are going to wrongly assume it's supposed to be updated whenever changes to the real data are made.  We've already seen exactly this problem in cdn-utils repo, e.g. see https://code.engineering.redhat.com/gerrit/#/c/198307/ where, as well as bumping the real version numbers used from within code, someone has wrongly thought it's necessary to update the testdata as well - because it looks the same.  I think that people just grep for the magic numbers/versions they know need to be updated and don't necessarily look at the docs first (or they assume the docs aren't complete).
- If the data is realistic then there is a risk that someone can develop code against it and not even realize the existence of the real data in a separate repo until problems occur.  e.g. imagine that someone writes some code which looks at the symlink_exceptions.  Then imagine we need to define a new exception, and we do that in cdn-definitions-private and adjust CDN accordingly.  Some time later this person files a bug against us stating "Hey, I've noticed this path doesn't use a symlink but you didn't list it in the exceptions!"  Then we have to explain that they should have deployed their tool/service to load from cdn-definitions-private if they wanted production data.  To be honest, I would even worry that whatever ticket master saw that complaint first is not going to realize they should tell the reporter to look at cdn-definitions-private, and they might instead waste time updating the field here similarly as in the gerrit change above.

In summary I worry that this data looking realistic is going to result in long-term confusion with people wrongly assuming that the realistic-looking data is *real* data.

Could we go through and force it to more obviously look like bogus/dummy data?  The idea would be that just about anyone looking at this file could quickly come to the conclusion "alright, this clearly isn't the true configuration used for production CDN right now, I need to go find where that's stored".

I get the impression this idea is already being applied somewhat as I note some differences here, e.g. the real RHUI ID isn't 222.  Can we push it further?

File: tests/test_schema.py [hunk 7]
```diff
@@ -45,3 +45,80 @@ def test_relative_path_not_match_schema(data, schema):
...
+def test_release_version_strings(data, schema, a, b):
+    """Validate the major_or_minor_version regex."""
+
+    # cfme_version_mappings are a major_or_minor_version to major_or_minor_version mapping.
+    data["cfme_version_mappings"]["platform_full_version"] == a
+    data["cfme_version_mappings"]["platform_version"] == b
+
+
+@pytest.mark.parametrize("major, minor", [("8.0", "8.0"), ("8", "8"), ("8.0", "8")])
+def test_release_version_strings(data, schema, major, minor):
```
Comment (rohanpm) [inline tests/test_schema.py:101]: This test is named identically to the previous one, this probably means only the second is being executed.

File: src/cdn_definitions/schema.yaml [hunk 8]
```diff
@@ -11,7 +11,211 @@ $id: http://release-engineering.github.io/cdn-definitions/schema.json
...
+
+  major_version:
+    type: string
+    pattern: "^[0-9]+$"
+
+  signing_key:
+    type: string
+    pattern: "^(37036783|A5787476|FD431D51|F21541EB|37017186|897DA07A|DB42A60E|\\[\\])$"
+
+  product_id:
```
Comment (rohanpm) [inline src/cdn_definitions/schema.yaml:35]: Should this be declared as an integer?

File: src/cdn_definitions/schema.yaml [hunk 9]
```diff
@@ -11,7 +11,211 @@ $id: http://release-engineering.github.io/cdn-definitions/schema.json
 
 
 definitions:
+  arch:
+    type: string
+    pattern: "^(x86|x86\\_64|ia64|i386|ppc|ppc64|ppc64le|s390|s390x|aarch64)$"
```
Comment (rohanpm) [inline src/cdn_definitions/schema.yaml:16]: I think this doesn't belong in the schema.  At first I thought it was borderline, but then I thought of this scenario:

"We're going to start shipping products for <new_arch>, this is company confidential until <some_marketing_event> on which date it will be both announced and released"

We would be forced to add the new secret arch into the public schema which wouldn't be acceptable.

Note: if you feel it's worthwhile to verify that arch values fall into some expected range, this could be an example of the kind of additional test beyond the schema which makes sense to add in cdn-definitions-private.

File: src/cdn_definitions/schema.yaml [hunk 10]
```diff
@@ -11,7 +11,211 @@ $id: http://release-engineering.github.io/cdn-definitions/schema.json
...
+    items:
+      $ref: "#/definitions/arch"
+
+  release_version:
+    type: string
+    pattern: "^[0-9]+\\.[0-9]+$"
+
+  major_version:
+    type: string
+    pattern: "^[0-9]+$"
```
Comment (rohanpm) [inline src/cdn_definitions/schema.yaml:29]: Should this be declared as an integer?

File: src/cdn_definitions/schema.yaml [hunk 11]
```diff
@@ -11,7 +11,211 @@ $id: http://release-engineering.github.io/cdn-definitions/schema.json
...
+      $ref: "#/definitions/major_version"
+    patternProperties:
+      ".*":
+        $ref: "#/definitions/release_version"
+    additionalProperties: false
+
+  release_stream_mapping:
+    type: object
+    properties:
+      aus:
```
Comment (rohanpm) [inline src/cdn_definitions/schema.yaml:85]: I think allowing arbitrary streams here might be a good idea, they do change sometimes and it might be desired to prepare for introducing a new stream before it's announced publicly.

File: src/cdn_definitions/schema.yaml [hunk 12]
```diff
@@ -11,7 +11,211 @@ $id: http://release-engineering.github.io/cdn-definitions/schema.json
...
+
+  # TPS stream configuration
+  rhel_ll_tps_stream_mapping:
+    type: object
+    propertyNames:
+      $ref: "#/definitions/release_version"
+    patternProperties:
+      ".*":
+        type: string
+        pattern: "^RHEL\\-\\%s\\-(LL|AUS)\\-\\%s$"
```
Comment (rohanpm) [inline src/cdn_definitions/schema.yaml:142]: I'm not sure the restriction here in the schema is right.  Looking at the actual data, it's always used the AUS style in RHEL6 and later.

I would expect the future is one of:

- Either: we *always* stick with the AUS style from here onwards.
   - In that case, this isn't too useful, the cdn-utils code may as well be "if RHEL5, then use LL format, otherwise use AUS format"
- Or: there is some case later on to introduce a new style
   - In that case, the schema as implemented now can't support it since it's restricted to only two possibilities, so you would have to go and update the schema to support a new value, then go and update your data to use exactly that value.

I'm guessing the former is more likely and I'd rather just not have this in the data.  I'm looking at `RHEL_LL_TPS_STREAM_MAPPING` and to me it seems that all the time spent updating this field across many release versions has been wasted, what with the value always being the same since RHEL6.  Some things would be better off *not* configurable if we want to encourage consistency going forward.

File: src/cdn_definitions/schema.yaml [hunk 13]
```diff
@@ -11,7 +11,211 @@ $id: http://release-engineering.github.io/cdn-definitions/schema.json
...
+    additionalProperties: false
+
+  uppercase_tps_variant_mappings:
```
Comment (rohanpm) [inline src/cdn_definitions/schema.yaml:145]: Could we perhaps not call this "uppercase"?

I'm looking at the related code now and I don't know why "upcase" is in the function name there, since it doesn't use uppercase values as inputs or outputs...

But anyway, even if the values were meant to be uppercase, I don't think that would need to go in the name unless there were some other non-uppercase mapping as well.

**Review by rohanpm — CHANGES_REQUESTED** [review]

File: src/cdn_definitions/schema.yaml (hunk 10 above)
Comment (crungehottman) [reply src/cdn_definitions/schema.yaml:31]: The jsonschema documentation states that object keys must be a string [1], and I created this definition to be used as an object key.  I will instead create a major_version_string definition (of type string) to be used solely as an object key/property name, and convert the major_version to type integer for all other uses (outside of a property name).

File: src/cdn_definitions/data.yaml (hunk 6 above)
Comment (crungehottman) [reply src/cdn_definitions/data.yaml:95]: +1. I think the data appears less realistic now.

File: src/cdn_definitions/schema.yaml (hunk 4 above)
Comment (crungehottman) [reply src/cdn_definitions/schema.yaml]: `[]` as a signing key is from `cdsutils.ProductContentSet.signatures`:
```
if self.data.platform == 'middleware':
    return []  # no RPM signing of zips
```

Comment (rohanpm) [reply src/cdn_definitions/schema.yaml]: Right, I thought `[]` might be something like that, but I don't think this is correct for reasons:

1. The underlying data (JSON or YAML) supports lists natively, so if we want to have an empty list as a value, we should just write an empty list into the file... not a list encoded into a string.  This is a case of "JSON within JSON" or "YAML within YAML" for no clear reason.
2. Is this field meant to represent a single signing key or a list of them?  It seems that it should be one or the other.  If a single signing key, then either null or an empty string is better to represent 'no signing key' than `[]`.

File: tests/test_schema.py [hunk 14]
```diff
@@ -45,3 +45,67 @@ def test_relative_path_not_match_schema(data, schema):
...
+
+    # cfme_version_mappings are a major_or_minor_version to major_or_minor_version mapping.
+    data["cfme_version_mappings"].append(
+        {"platform_full_version": a, "platform_version": b}
+    )
+    jsonschema.validate(data, schema)
+
+
+# @pytest.mark.parametrize("major, minor", [("8.0", "8.0"), ("8", "8"), ("8.0", "8")])
+# def test_release_version_strings(data, schema, major, minor):
```
Comment (rohanpm) [inline tests/test_schema.py:88]: What's the deal with this, can we get it re-enabled or removed?

**Review by nathanegillett — APPROVED** [review]

File: src/cdn_definitions/schema.json [hunk 15]
```diff
@@ -7,6 +7,124 @@
...
+                }
+            },
+            "propertyNames": {
+                "pattern": "^(current|latest)\\_rhel[0-9]+\\_(beta|dist)$",
+                "type": "string"
+            },
+            "type": "object"
+        },
+        "env_to_releasever_mapping": {
+            "additionalProperties": false,
```
Comment (rohanpm) [inline src/cdn_definitions/schema.json:68]: Hmm, I don't think this is correct because we do actually make it possible right now for the user of cdn-utils to define their own environments, so "qa", "stage", "prod" aren't the only values.  Search code for /etc/rcm/cdn-environments.json.  I think the ability should be retained - it's preferable for testers to e.g. define their own test1, test2 environment rather than having to *overwrite* one of the existing environments if they want to test against their own infrastructure.

Can we make this one a patternProperties ".*" to allow arbitrary environments?  "required" qa, stage, prod may still make sense though.

File: src/cdn_definitions/schema.json [hunk 16]
```diff
@@ -29,15 +147,169 @@
...
+        },
+        "release_version_list": {
+            "items": {
+                "$ref": "#/definitions/release_version"
+            },
+            "type": "array",
+            "uniqueItems": true
+        },
+        "signing_key": {
+            "pattern": "^[A-Fa-f0-9]{8}",
```
Comment (rohanpm) [inline src/cdn_definitions/schema.json:203]: I don't think we should support both A-F and a-f, can we pick one or the other?

I wasn't sure about it but then I saw below in signing_key_list that you put uniqueItems: true.  That doesn't provide much value if, for example, "abcdef00" and "ABCDEF00" are both allowed to be in the list.

File: src/cdn_definitions/schema.yaml [hunk 17]
```diff
@@ -11,7 +11,192 @@ $id: http://release-engineering.github.io/cdn-definitions/schema.json
...
+        $ref: "#/definitions/release_version_list"
+      e4s:
+        $ref: "#/definitions/release_version_list"
+    requiredProperties:
+    - aus
+    - eus
+    - tus
+    - e4s
+    patternProperties:
+      ".*":
```
Comment (rohanpm) [inline src/cdn_definitions/schema.yaml:101]: A question, since this .* will already match all the existing eus, tus etc, is it necessary to list those explicitly?  i.e. if you just delete the "properties" block, does it already work with same semantics.

**Review by rohanpm — CHANGES_REQUESTED** [review]

File: src/cdn_definitions/schema.json (hunk 15 above)
Comment (crungehottman) [reply src/cdn_definitions/schema.json:68]: +1, populating the prod, stage, and qa mappings would be additional work for testers, so I'll also remove the envs from the required properties list. This sort of requirement could instead be enforced via a test in cdn-defintions-private. I'll leave the envs in data.yaml as an example.

File: src/cdn_definitions/schema.yaml (hunk 17 above)
Comment (crungehottman) [reply src/cdn_definitions/schema.yaml:101]: No, it seems that I do not need to list them explicitly.

I have reconsidered explicitly requiring any of the streams. This is another requirement that could instead be enforced via a test in cdn-defintions-private. If we were to add an additional stream to cdn-definitions-private and wanted to avoid disclosing this information to the public, we'd then have to worry about adding that stream to the "required" list of update streams in cdn-definitions after the release was announced publicly. I think a test would serve us just as well, and would be less annoying to update.

I'll leave the four streams in data.yaml as an example, but I will remove the streams from the list of required properties.

File: src/cdn_definitions/schema.yaml [hunk 18]
```diff
@@ -11,7 +11,170 @@ $id: http://release-engineering.github.io/cdn-definitions/schema.json
...
+  major_version:
+    type: integer
+
+  major_version_string:
+    type: string
+    pattern: "^[0-9]+$"
+
+  signing_key:
+    type: string
+    pattern: "^[A-F0-9]{8}"
```
Comment (rohanpm) [inline src/cdn_definitions/schema.yaml:36]: Just noticed this is missing the ending anchor "$", so after the initial 8 characters it would accept any old garbage.

**Review by rohanpm — CHANGES_REQUESTED** [review]
Are there any known reasons remaining not to submit & release this as soon as it's approved?

**Comment by crungehottman** [comment]
> Are there any known reasons remaining not to submit & release this as soon as it's approved?

No, please submit and release this as soon as it's approved.

**Review by rohanpm — APPROVED** [review]

**Comment by rohanpm** [comment]
About the release of this, I want to look into doc updates first and try to have an accurate changelog and updated docs first (or at least any wrong info removed).  But if that looks to be too much work, I won't worry about it.
